File: 2024 August/Sum Tree/main.py
# Link: https://www.geeksforgeeks.org/problems/sum-tree/1

from collections import deque


class Node:
    def __init__(self, val):
        self.data = val
        self.left = None
        self.right = None


def new_node(val):
    return Node(val)


def build_tree(s):
    # Corner Case
    if not s or s[0] == "N":
        return None

    # Creating list of strings from input string after splitting by space
    ip = s.split()

    # Create the root of the tree
    root = new_node(int(ip[0]))

    # Push the root to the queue
    queue = []
    queue.append(root)

    # Starting from the second element
    i = 1
    while queue and i < len(ip):
        # Get and remove the front of the queue
        curr_node = queue.pop(0)

        # Get the current node's value from the string
        curr_val = ip[i]

        # If the left child is not null
        if curr_val != "N":
            # Create the left child for the current node
            curr_node.left = new_node(int(curr_val))

            # Push it to the queue
            queue.append(curr_node.left)

        # For the right child
        i += 1
        if i >= len(ip):
            break
        curr_val = ip[i]

        # If the right child is not null
        if curr_val != "N":
            # Create the right child for the current node
            curr_node.right = new_node(int(curr_val))

            # Push it to the queue
            queue.append(curr_node.right)
        i += 1

    return root


class Solution:
    def check_three_sum(self, root, ans):
        if root.left is None and root.right is None:
            return root.data

        sum_left = 0
        sum_right = 0
        if root.right:
            sum_left = self.check_three_sum(root.right, ans)

        if root.left:
            sum_right = self.check_three_sum(root.left, ans)

        if root.data != sum_left + sum_right:
            ans[0] = False

        total_sum = root.data + sum_left + sum_right
        return total_sum

    def is_sum_tree(self, node):
        ans = [True]
        self.check_three_sum(node, ans)
        return ans[0]


if __name__ == "__main__":
    obj = Solution()
    s = "3 1 2"
    root = build_tree(s)
    print(obj.is_sum_tree(root))

    s = "10 20 30 10 10"
    root = build_tree(s)
    print(obj.is_sum_tree(root))

File: 2024 August/Sum Tree/main.cpp
// Link: https://www.geeksforgeeks.org/problems/sum-tree/1

#include <iostream>
#include <queue>
#include <sstream>
#include <vector>
using namespace std;

struct Node
{
    int data;
    struct Node *left;
    struct Node *right;
};

// Utility function to create a new Tree Node
Node *newNode(int val)
{
    Node *temp = new Node;
    temp->data = val;
    temp->left = NULL;
    temp->right = NULL;

    return temp;
}

// Function to Build Tree
Node *buildTree(string str)
{
    // Corner Case
    if (str.length() == 0 || str[0] == 'N')
        return NULL;

    // Creating vector of strings from input
    // string after spliting by space
    vector<string> ip;

    istringstream iss(str);
    for (string str; iss >> str;)
        ip.push_back(str);

    // Create the root of the tree
    Node *root = newNode(stoi(ip[0]));

    // Push the root to the queue
    queue<Node *> queue;
    queue.push(root);

    // Starting from the second element
    int i = 1;
    while (!queue.empty() && i < ip.size())
    {

        // Get and remove the front of the queue
        Node *currNode = queue.front();
        queue.pop();

        // Get the current node's value from the string
        string currVal = ip[i];

        // If the left child is not null
        if (currVal != "N")
        {

            // Create the left child for the current node
            currNode->left = newNode(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->left);
        }

        // For the right child
        i++;
        if (i >= ip.size())
            break;
        currVal = ip[i];

        // If the right child is not null
        if (currVal != "N")
        {

            // Create the right child for the current node
            currNode->right = newNode(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->right);
        }
        i++;
    }

    return root;
}

class Solution
{
public:
    int checkThreeSum(Node *root, bool *ans)
    {
        if (root->right == NULL && root->left == NULL)
            return root->data;

        int sumleft = 0;
        int sumright = 0;
        if (root->right)
            sumleft = checkThreeSum(root->right, ans);

        if (root->left)
            sumright = checkThreeSum(root->left, ans);

        if (root->data != sumleft + sumright)
            *ans = false;

        int totalsum = root->data + sumleft + sumright;
        return totalsum;
    }

    bool isSumTree(Node *root)
    {
        bool ans = true;
        checkThreeSum(root, &ans);
        return ans;
    }
};

int main()
{
    Solution obj;
    string s = "3 1 2";
    Node *root = buildTree(s);
    cout << obj.isSumTree(root) << endl;

    s = "10 20 30 10 10";
    root = buildTree(s);
    cout << obj.isSumTree(root) << endl;

    return 0;
}